js_modules/dagster-ui/packages/ui-core/src/hooks/useQueryAndLocalStoragePersistedState.tsx

```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+
+import {QueryPersistedDataType, useQueryPersistedState} from './useQueryPersistedState';
+import {useSetStateUpdateCallback} from './useSetStateUpdateCallback';
+
+/**
+ *
+ * Use URL query string as main source of truth with localStorage as the backup if no state is found in the query string
+ * Syncs changes back to localStorage but relies solely on queryString after the initial render
+ * @returns
+ */
+export const useQueryAndLocalStoragePersistedState = <T extends QueryPersistedDataType>(
+  props: Parameters<typeof useQueryPersistedState<T>>[0] & {
+    localStorageKey: string;
+    isEmptyState: (state: T) => boolean;
+  },
+): [T, (setterOrState: React.SetStateAction<T>) => void] => {
+  // Grab state from localStorage as "initialState"
```

Comment:
Hmm could this use `useStateWithStorage(props.localStorageKey, props.decode || (v) => v)`? I think the decode method could be the validate method passed to that hook. useStateWithStorage has a couple bonus features we don't need I guess?